pipeline/data_store.py

```diff
@@ -0,0 +1,773 @@
+                    prov = [p for p in upstream_provs if p.process == name][0]
+
+                # second, try to get a provenance from objects saved to the store:
+                elif obj is not None and hasattr(obj, 'provenance') and obj.provenance is not None:
+                    prov = obj.provenance
+
+                # last, try to get the latest provenance from the database:
+                else:
+                    prov = get_latest_provenance(name, session=session)
+
+                # can't find any provenance upstream, therefore
+                # there can't be any provenance for this process
+                if prov is None:
+                    return None
+
+                upstreams.append(prov)
+
+            # we have a code version object and upstreams, we can make a provenance
+            prov = Provenance(
+                process=process,
+                code_version=code_hash.code_version,
+                parameters=pars_dict,
+                upstreams=upstreams,
+            )
+            prov.update_hash()  # need a new object to calculate the hash, then check if it exists on the DB:
+            existing_p = session.scalars(
+                sa.select(Provenance).where(
+                    Provenance.unique_hash == prov.unique_hash
+                )
+            ).first()
+
+            if existing_p is not None:
+                prov = existing_p
+
+        return prov
+
+    def _get_provenance_fallback(self, process, session=None):
+        """
+        Get the provenance for a given process, without knowing
+        the parameters or code version.
+        This simply looks for a matching provenance in the upstream_provs
+        attribute, and if it is not there, it will call the latest provenance
+        from the database.
+
+        This will raise if no provenance can be found.
+        """
+        # see if it is in the upstream_provs
+        prov_list = [p for p in self.upstream_provs if p.process == process]
+        provenance = prov_list[0] if len(prov_list) > 0 else None
+
+        # try getting the latest from the database
+        if provenance is None:  # check latest provenance
+            provenance = get_latest_provenance(process, session=session)
+        if provenance is None:  # still can't find anything!
+            raise ValueError(f'Cannot find the "{process}" provenance!')
+
+        return provenance
+
+    def get_raw_exposure(self, session=None):
+        """
+        Get the raw exposure from the database.
+        """
+        if self.exposure is None:
+            if self.exp_id is None:
+                raise ValueError('Cannot get raw exposure without exp_id!')
+
+            with SmartSession(session) as session:
+                self.exposure = session.scalars(sa.select(Exposure).where(Exposure.id == self.exp_id)).first()
+
+        return self.exposure
+
+    def get_image(self, provenance=None, session=None):
```

Comment:
Unless I missed it, you don't have the code here to load the actual pixel values, just the database structure.

Is the plan eventually to have code in here that secures the actual image file in local storage, and loads the actual pixel values into memory?  (The way I do it in lensgrinder, that wouldn't happen here, but the image structure itself has the information necessary for another routine to find the fits file on local storage and read it, but I know that you wanted to just keep the image in memory all the time.)

(This matters because of my earlier comment about not storing the pixel values for the entire exposure.)

Reply:
The way this will work is that the Exposure and Image classes would have, besides the SQLA interface to their respective rows in the DB, also some code to lazy load the data. 

I thought about something like `self._data` and `self.data` and when you ask for `self.data` it checks if `self._data` is populated and if not it fetches it from disk. 
There would be a "clear data" method that the user can opt in to calling when memory conservation is an issue.